File: csharp/components/Strata.Application/Service/AbstractService.cs
//  ##########################################################################
//  # File Name: AbstractService.cs
//  # Copyright: 2018, Sapientia Systems, LLC.
//  ##########################################################################

using Strata.Foundation.Logging;
using Strata.Domain.UnitOfWork;
using Strata.Application.Interception;
using Strata.Application.Messaging;

namespace Strata.Application.Service
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    public abstract
    class AbstractService:
        IUnitOfWorkPropertySupplier
    {
        public IUnitOfWorkProvider Provider { get; set; }
        public IActionQueue        Queue { get; set; }
        public ILogger             Logger { get; set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractService(
            IUnitOfWorkProvider provider,
            ILogger             logger):
            this(provider,null,logger) {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractService(
            IUnitOfWorkProvider provider,
            IActionQueue        queue,
            ILogger             logger)
        {
            Provider = provider;
            Queue    = queue;
            Logger   = logger;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected void
        PushRollbackAction(RollbackAction rollbackAction)
        {
            Provider
                .GetUnitOfWork()
                .PushRollbackAction(rollbackAction);
        }
    }

}

//  ##########################################################################

File: java/Components/Strata.Application.Core/src/strata/application/core/service/AbstractService.java
//////////////////////////////////////////////////////////////////////////////
// AbstractService.java
//////////////////////////////////////////////////////////////////////////////

package strata.application.core.service;

import strata.application.core.interception.IUnitOfWorkPropertySupplier;
import strata.domain.core.unitofwork.IUnitOfWorkProvider;
import strata.foundation.core.action.IActionQueue;

import static strata.foundation.core.utility.Awaiter.await;

public abstract
class AbstractService
    implements IUnitOfWorkPropertySupplier
{
    private final IUnitOfWorkProvider provider;
    private final IActionQueue        queue;

    protected
    AbstractService(IUnitOfWorkProvider p)
    {
        this(p,null);
    }

    protected
    AbstractService(
        IUnitOfWorkProvider p,
        IActionQueue        q)
    {
        provider = p;
        queue    = q;
    }

    @Override
    public IUnitOfWorkProvider
    getProvider()
    {
        return provider;
    }

    @Override
    public IActionQueue
    getQueue()
    {
        return queue;
    }

    protected void
    pushRollbackAction(Runnable action)
    {
        await(provider.getUnitOfWork()).pushRollbackAction(action);
    }
}

//////////////////////////////////////////////////////////////////////////////
